ostree: new option for the overlay driver

giuseppe opened the pull request:
if `ostree_repo` is specified, then any layer is first committed to the specified ostree repository and then checkout out to the same location so that files are deduplicated using OSTree hardlinks.

The `exclude_ostree` BUILDTAG can be used to disable ostree support.

giuseppe commented:
@rhatdan, I've pushed another version with all the comments addressed ⬆️

rhatdan commented on drivers/ostree/driver.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+func getDir(home, id string) string {
+	return fmt.Sprintf("%s/%s/diff", home, id)
```
We tend to use filepath.Join() for this type of thing.

rhatdan commented on drivers/ostree/driver.go:
```diff
@@ -0,0 +1,178 @@
+// Create prepares the filesystem for the OSTREE driver and copies the directory for the given id under the parent.
+func (d *Driver) convertToOSTree(root, id string) error {
+	if _, err := d.repo.PrepareTransaction(); err != nil {
+		return fmt.Errorf("Could not prepare the OSTree transaction: %v", err)
```
Use 
errors.Wrapf(err, "could not prepare the OSTree transaction")

Standard is to have errors start with lowercase. I believe.

rhatdan commented on drivers/ostree/driver.go:
```diff
@@ -0,0 +1,178 @@
+	branch := fmt.Sprintf("ocilayer/%s", id)
+
+	if _, err := d.repo.Commit(root, branch, commitOpts); err != nil {
+		return fmt.Errorf("Could not commit the layer: %v", err)
```
errors.Wrapf.

rhatdan replied:
lowercase

rhatdan commented on drivers/ostree/driver.go:
```diff
@@ -0,0 +1,178 @@
+	}
+
+	if _, err := d.repo.CommitTransaction(); err != nil {
+		return fmt.Errorf("Could not complete the OSTree transaction: %v", err)
```
Ditto

rhatdan commented on drivers/ostree/driver.go:
```diff
@@ -0,0 +1,178 @@
+	}
+
+	if err := system.EnsureRemoveAll(root); err != nil {
+		return err
```
errors.Wrapf(err, "failed to remove %q", root)

giuseppe replied:
I looked at the code and it already gives a similar error message, so I think it would be duplicated `return errors.Wrapf(e, "error while removing %s", dir)`.  In some other cases it returns directly the error from `os.RemoveAll()`

rhatdan commented on drivers/ostree/driver.go:
```diff
@@ -0,0 +1,178 @@
+
+	checkoutOpts := otbuiltin.NewCheckoutOptions()
+	if err := otbuiltin.Checkout(d.repoLocation, root, branch, checkoutOpts); err != nil {
+		return fmt.Errorf("Could not checkout from OSTree: %v", err)
```
errors.Wrapf

rhatdan commented on drivers/ostree/driver.go:
```diff
@@ -0,0 +1,178 @@
+func (d *Driver) Remove(id string) error {
+	branch := fmt.Sprintf("ocilayer/%s", id)
+	if err := exec.Command("ostree", fmt.Sprintf("--repo=%s", d.repoLocation), "refs", "--delete", branch).Run(); err != nil {
+		return fmt.Errorf("Could not delete OSTree branch: %v", err)
```
errors.Wrapf("Could not delete OSTree branch")

rhatdan commented on drivers/ostree/driver.go:
```diff
@@ -0,0 +1,178 @@
+
+// Get returns the directory for the given id.
+func (d *Driver) Get(id, mountLabel string) (string, error) {
+	path, err := d.overlay.Get(id, mountLabel)
```
return d.overlay.Get(id, mountLabel)

rhatdan commented on drivers/ostree/driver_stub.go:
```diff
@@ -0,0 +1,18 @@
+}
+
+func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (graphdriver.Driver, error) {
+	return nil, fmt.Errorf("ostree driver not supported")
```
Use errors.Errorf()

giuseppe commented:
thanks for the quick review.  I've addressed the new comments

rhatdan commented:
@nalind Should we add a new driver or enhance the Overlay driver to use OSTREE via an option?  
overlay.

giuseppe commented on drivers/ostree/driver.go:
```diff
@@ -0,0 +1,182 @@
+	if convert {
+		err := d.convertToOSTree(root, id)
+		if err != nil {
+			return err
```
yes, it is defined in this module and it already gives useful errors

rhatdan replied:
Thanks for checking.

giuseppe commented:
I've changed it to be an option for overlay instead of a new driver

rhatdan commented:
I find your description of the PR to be very confusing, you talk about skopeo requiring a patch that is different then this patch set.  Can you fix up the commit message for the patch.

Also you are using a different BUILDTAG for elimination of ostree then containers/image uses. Wouldn't we need to use both BUILDTAGS to to not require OSTREE?

giuseppe commented:
sure, I forgot to update the PR after I changed the implementation to be an option instead of a new driver.

rhatdan commented:
What about `exclude_ostree` versus `containers_image_ostree_stub`

giuseppe commented:
should I use `containers_image_ostree_stub`?  That name sounds specific to containers/image, that is why I've used exclude_ostree, but the end result is the same

rhatdan commented:
I hate the name containers_image_ostree_stub and much prefer exclude_ostree. but I think to get this to work we would need to specify both?  @mitrmac WDYT
Could we change the boolean in containers/image to exclude_ostree or are we too late?  I guess we could add it to containers/image and support both?

nalind commented:
How much space are we typically saving when enabling this?

giuseppe commented:
I've done only a quick test:

```
## with overlay
# for i in gscrivano/etcd gscrivano/flannel; do \
  skopeo copy docker://$i containers-storage:\[overlay@/var/lib/containers/storage+/var/run/containers/storage]$i
  done
# du --summarize -x --si /var/lib/containers/storage/
406M        /var/lib/containers/storage/

## with ostree
# for i in gscrivano/etcd gscrivano/flannel; do \
  skopeo copy docker://$i containers-storage:\[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay2.ostree_repo=/var/lib/containers/storage/overlay2/ostree/.repo]$i
  done
# du --summarize -x --si /var/lib/containers/storage/
340M        /var/lib/containers/storage/
```

rhatdan commented:
16% savings, but the real question would be if you had 100 images.

giuseppe commented:
yes, it works better with more images, the more images you have the bigger chance to have the same files coming from different layers.

I've tried with a bunch of images, this is the script I've used:

https://gist.github.com/giuseppe/95f917fa271bcdc6555851cc4774d83a

Please be careful that it wipes the entire overlay storage, edit the path to something else if you care to keep it.

This is what I got:

```
3.2G	/var/lib/containers/storage
2.1G	/var/lib/containers/storage-ostree/
```

rhatdan commented:
@mtrmac Would this cause verification to fail since timestamps will end up different?

rhatdan commented:
@vbatts ^^

mtrmac commented:
> @mtrmac Would this cause verification to fail since timestamps will end up different?

(Disclaimer: I really don’t know all that much about c/storage.) Looking at https://github.com/containers/storage/blob/master/layers.go , `layerStore.ApplyDiff` and `layerStore.Diff` use tar-split at the general level, before invoking the underlying drivers to record the actual files. IIUC tar-split stores the raw tar headers, i.e. including the timestamps, so the filesystem timestamps should not matter — at least as far as c/image and extracting the image is concerned, at least. If the on-filesystem timestamps did not match the original layers, this would probably be visible, and detectable, from within a container.

rhatdan commented:
@vbatts Could you confirm whether this is going to be a problem?

vbatts commented:
@rhatdan as long as it is an opt-in basis. The C linkage to libostree makes this difficult for minikube, and I'm likely openshift too

vbatts commented:
oh, for tar-split ... lemme think

rhatdan commented:
Yes you will be able to compile it out.

vbatts commented:
for the re-assembly of a layer it would not use inode attributes of the files, only needing the path to the file and the contents of the file to be intact.
If you build an entirely new layer and export it, then the initial tar pushed would use the mtimes of the filesystem at the time of push (producing a tar object).

I don't see any technical issues, though users expectation of mtimes may be subject to confusion. 

If there were a thin copy-up for attribute changes, then this could be a case for mtree (https://github.com/vbatts/go-mtree) so the prepared rootfs could have it's attributes like mtime restored from what the tar archive described.

giuseppe commented:
I've started working on OSTree to allow deduplication of files with different SELinux labels using reflinks where possible, such as XFS on new kernels:

With XFS reflinks we will have different inodes, so if needed we can setup different attributes for a file even if it is sharing the payload with a different file.

https://github.com/ostreedev/ostree/pull/1443

This is only a first step, I will also need to:

- change the lookup in OSTree to find the same files in parent repositories
- drop the bind mount on `/var/lib/containers/storage/[driver]`

This will allow to share files with OSTree on the host, and system containers.

This is a local optimization, it doesn't change the way images are pulled or stored on the server/registry.

cgwalters commented:
There are some upsides and downsides to this that I see.  Space savings is good, but OTOH right now libostree's locking story isn't fully fleshed out; see https://github.com/ostreedev/ostree/pull/1292 

And even if we had the locking fully done, this would mean installing/removing containers could block/slow host updates.  Of course we could prioritize host updates.  Another issue related to this is that personally right now I treat my container storage as fully disposable.  If it's entangled with the host storage, that gets harder.

I feel like we should be focusing on the system container storage path first; those usually change with the host, and are critical for bootstrap (i.e. they're distinct from app containers).

That all said this is a really interesting space to explore for sure.

giuseppe commented:
I agree with you.  I think we should still have a separate ostree repository for the app containers.  It is important to be able to easily wipe the entire containers storage if required without affecting the rest of the system.

alexlarsson commented:
Could we use ostree for the store, and then rofiles-fuse --copyup instead of overlayfs for deduplicated COW non-root support?

giuseppe commented:
@alexlarsson I guess so, but wouldn't that be a completely different storage driver?

After the discussion with running Buildah as non-root user, I agree that an `ostree` storage could be a very useful one, and much better than my first version of this patchset where I was adding a new `ostree` storage driver that was a clone of overlay+ostree storage.

alexlarsson commented:
Yeah, a pure ostree storage driver seem useful. There are some possible issues with it not preserving mtimes though.

rhatdan commented:
Sounds good, my only question would be do we want both.  This one for deduplication of resources, so less storage is used with overlay and the other for the non-privileged use case.  Not sure preservice mtimes is a big deal when building container images.

giuseppe commented:
I think we want both.  The pure ostree storage driver is useful in the rootless case, while having deduplication with overlay is another feature.

Is there anything blocking this PR?

rhatdan commented:
Did we ever fix up the ostree buildtag issues?

giuseppe commented:
I changed it to `exclude_ostree`, is that fine or should I use a different buildtag?

rhatdan commented:
No that is fine, but I want to get the same buildatag into containers/image  So we have consistency.

mtrmac commented:
c/image currently uses `containers_image_ostree_stub`.  Consistency in Go tags is just another way to say “namespace collision” ;-), though I do see how it is attractive for consumers of multiple libraries.

rhatdan commented:
Right can't we add exclude_ostree also to containers/image, so if the user of the library sets that flag it will not get compiled in.  Although I would think we want a positive instead of a negative. Like just ostree, This is what we do with SELinux and Apparmor.  If you specify a buildahtag og ostree you get the bindings, if you don't you don't get them.

mtrmac commented:
\*shrug\* I can live with that.

rhatdan commented:
@giuseppe Could you reverse this so ostree and !ostree are the build tags to match selinux, apparmor etc...

giuseppe commented:
I've changed "exclude_ostree" to "!ostree" :arrow_up:

giuseppe commented:
I've added a new option for skipping the bind mount on `/var/lib/containers/storage/overlay*`, so that the ostree repository can be placed outside of that directory, and more important: we can use an ostree parent repository for deduplicating files in two different storages.

rhatdan commented on .travis.yml:
```diff
@@ -5,10 +5,12 @@ go:
 dist: trusty
 sudo: required
+env:
+  - TAGS='!ostree'
```
Shouldn't just not having TAGS defined make this work?

giuseppe replied:
yes true, I'll drop that

rhatdan commented:
You need to add code to Makefile ALLTAGS to discover if it should use BUILDTAG=ostree.

Something like
ls -l hack/*tag.sh
-rwxrwxr-x. 1 dwalsh dwalsh 132 Sep 29 15:32 hack/btrfs_tag.sh
-rwxrwxr-x. 1 dwalsh dwalsh 332 Sep 29 15:32 hack/libdm_tag.sh

giuseppe commented:
changes done and tests pass again :arrow_up:

rhatdan commented:
LGTM

nalind commented on drivers/driver_linux.go:
```diff
@@ -59,6 +59,7 @@ var (
 		"btrfs",
 		"zfs",
 		"vfs",
+		"ostree",
```
There's no call to the `Register()` function with this name, so I don't think this does anything.

giuseppe replied:
this chunk must be deleted as now I am not adding a new driver but only an option for the overlay driver.

Thanks for spotting it!

nalind commented on drivers/overlay/ostree.go:
```diff
@@ -0,0 +1,122 @@
+
+	checkoutOpts := otbuiltin.NewCheckoutOptions()
+	checkoutOpts.RequireHardlinks = true
+	checkoutOpts.Whiteouts = true
```
Whiteout information is stored differently for overlay than it is by aufs, and the format is converted when creating and extracting archives using pkg/archives.  What does this flag cause ostree to do?

giuseppe replied:
it tells ostree to handle files with a prefix ".wh." (i.e. delete the existing file/directory) in the checkout directory.

nalind replied:
That's how aufs stores whiteouts (and I'm not sure that what you're describing handles opaque directories), but in overlay whiteouts are character device nodes with major/minor of 0/0.  (Under [whiteouts and opaque directories](https://www.kernel.org/doc/Documentation/filesystems/overlayfs.txt) in the overlayfs docs.)

nalind replied:
@cgwalters Does ostree handle opaque directory whiteouts?  I took a quick look at `ostree-repo-checkout.c` and I couldn't find logic specific to handling entries named `.wh..wh..opq`.

nalind replied:
@cgwalters It appears not, then.

giuseppe replied:
I've opened a PR to support this: https://github.com/ostreedev/ostree/pull/1486

nalind replied:
Thanks for that.
If I'm reading it right, `fixFiles` removes file whiteouts from diff directories, and that's a problem.

giuseppe replied:
ah thanks to have found this issue.  I thought there were no whiteouts files, I've changed the code so that there is no special handling of these files.

nalind commented:
Is there a particular rationale for calling out to the ostree CLI in `createOSTreeRepository()` and `deleteOSTree()`, but using the `otbuiltins` package in `convertToOSTree()`?

nalind commented on drivers/overlay/ostree.go:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	return nil
+}
```
This function might be simpler if it used `path/filepath.Walk()`.

giuseppe replied:
done in the new version

giuseppe commented:
> Is there a particular rationale for calling out to the ostree CLI in createOSTreeRepository() and deleteOSTree(), but using the otbuiltins package in convertToOSTree()?

these functionalities are not exposed by the Golang bindings yet :-)

nalind commented on drivers/overlay/ostree.go:
```diff
@@ -0,0 +1,110 @@
+			if err := os.Remove(path); err != nil {
+				return err
+			}
+			if info.IsDir() {
```
This looks like it's nested at the wrong level - a directory is unlikely to be a pipe/socket/device.

nalind commented on drivers/overlay/ostree.go:
```diff
@@ -0,0 +1,110 @@
+						return err
+					}
+				}
+			} else if usermode && (info.Mode().IsRegular()) {
```
This also looks like it's nested at the wrong level.

rhatdan commented:
@giuseppe Could you setup a WIP vendor of this into Buildah or Podman and attempt to remove some files and rename a directory and see if everything works correctly.

giuseppe commented:
I've tried a custom build for podman, I needed to manually workaround some build issues caused by the revendoring (not caused by containers/storage, in facts I can see the same just by running `vndr` on origin master).

This change is required in podman to start using ostree:
```
--- a/hack/ostree_tag.sh
+++ b/hack/ostree_tag.sh
@@ -1,4 +1,6 @@
 #!/bin/bash
-if ! pkg-config ostree-1 2> /dev/null ; then
+if pkg-config ostree-1 2> /dev/null ; then
+       echo ostree
+else
        echo containers_image_ostree_stub
 fi
```

and this is what I have in my /etc/containers/storage.conf:
```

ostree_repo = "/ostree/repo"

skip_mount_home = "true"
```

after that everything in my limited testing works as expected in the container:

```
/usr # cd /
/ # ls -l
total 56
drwxr-xr-x    2 root     root         12288 Jan  1  1970 bin
drwxr-xr-x    5 root     root           360 Mar 14 17:50 dev
drwxr-xr-x    1 root     root          4096 Mar 14 17:50 etc
drwxr-xr-x    2 root     root          4096 Mar 14 17:51 foo
drwxr-xr-x    2 nobody   nogroup       4096 Jan  1  1970 home
dr-xr-xr-x  340 root     root             0 Mar 14 17:50 proc
drwx------    1 root     root          4096 Mar 14 17:50 root
dr-xr-xr-x   13 root     root             0 Mar 14 11:23 sys
drwxrwxrwt    2 root     root          4096 Jan  1  1970 tmp
drwxr-xr-x    1 root     root          4096 Mar 14 17:51 usr
drwxr-xr-x    4 root     root          4096 Jan  1  1970 var
/ # touch /usr/foo
/ # rm /usr/foo
/usr # mkdir bar
/usr # touch bar/file
/usr # mv bar baz
/usr # rm -rf baz
```

nalind commented:
The main thing I'd want to be sure of is that we're handling whiteout information correctly, since overlay uses device nodes for whiteout, and extended attributes to mark opacity.  Specifically, I'd want to be sure that if an image creates files or directories in one layer, then removes them in a second layer that is applied on top of that first layer, that we're not losing that information when we extract the image into local storage.  Likewise, if a directory is marked as opaque, that the commit/checkout step doesn't discard that information.

One way to verify this would be to find a multi-layered image whose history shows it creating files in one layer, then using "rm" to remove those files in a different layer, creating whiteout information in that layer, or an image where a directory was marked as opaque, signifying that it had been removed and then recreated within a single layer.  (That an image does this could be verified for certain by copying the image to a "dir:" location with skopeo, and examining the layer blobs directly using "tar tvf".  Whiteout entries for files and directories will have `.wh.` prefixed to their basenames, and opaque directories contain a `.wh..opq.` "file".)

For example, `docker.io/ubuntu@sha256:d3fdf5b1f8e8a155c17d5786280af1f5a04c10e95145a515279cf17abdf0191f` contains a layer with blobsum `3009ec50c8871f7a3f19c248194270a1a3aa28705934d619acbd96c663ce5c89` which removes the files in `/var/lib/apt/lists/`, while `docker.io/redis@sha256:26c93c5b06eaa323bb1089500f42b0dd158138772348b865e364127f1d554982` contains a layer with blobsum `67b22db27e5144c14efdf55d9cef9c3dadffb43593790ec99663fb1b783dbb6b` which marks `/var/lib/apt/lists/partial` as opaque.

The root filesystem of a working container created by versions of buildah built with and without the patch, if both versions are forced to download the image, should have the same set of files, with the same contents.  In particular, deleted files for which whiteout exists in the image's blobs, and the contents of directories from layers before they were marked as opaque, should not be present in the container.

giuseppe commented:
@nalind I had  to do one change in the `fixFiles` function.  When we encounter a file that ostree does not know how to manage (as a block device) we skip the ostree deduplication for the layer.  I needed to do this to support the 0/0 character devices whiteout.

We could add support for these files later on, but first it requires some changes in the ostree golang bindings otherwise we need to handle the entire logic of pruning and recreating these files.

With the new version I've pushed the whiteout layers are handled correctly and the deleted files are not visible from the container.

giuseppe commented:
@nalind does this look fine now?

rhatdan commented:
@giuseppe Lets rebase and get this merged.

giuseppe commented:
bot, retest this please

rhatdan commented:
LGTM

rhatdan commented:
@nalind PTAL

giuseppe commented:
I've rebased it again.  Is it useful to have this feature or should I close this PR?

rhatdan commented:
I believe this is useful.  As long as we make it easy to compile out.

giuseppe commented:
@rhatdan @nalind it is possible to control if OSTree is built or not.  I've verified that whiteout device files are kept

giuseppe commented:
@nalind @rhatdan I've added a test for the ostree deduplication part.  I've added it to vfs as well so that can be tested on Travis as well.  I've tried it on overlay locally using a Vagrant VM

rhatdan commented:
@giuseppe Tests are failing.

giuseppe commented:
@rhatdan fixed.  I needed to set the mtime to the EPOCH so it doesn't trigger a difference when OSTree is used.

giuseppe commented:
the only missing part that I would like to work on as a follow-up is dropping the `ostree` CLI usage and add the missing features to the go bindings (delete branch and init repository).

rhatdan commented:
LGTM
@nalind PTAL

giuseppe commented:
rebased

giuseppe commented:
@rhatdan fine to merge this?

rhatdan commented:
Going to finally merge this one.
